### PR TITLE
Add service support

### PR DESCRIPTION
**Pull request by SamVerschueren**
Created a first version of the service support. The only service supported at the moment is `mongodb`. If we think the implementation is good, I will add support for others as well.

This PR is not directly meant to be merged, rather to provide feedback and think together on how we can improve this.

**note 1**: `travis` exposes [mongodb](http://docs.travis-ci.com/user/database-setup/#MongoDB) at the `127.0.0.1`. This seems not possible when linking containers, but not sure, maybe there is a way. So at the moment, mongodb is exposed as `db` host.

**note 2**: I used the `latest` version of `mongodb`, but I believe `travis` is exposes `2.4.x` or something (which is a shame off course). We should definitely use the same version as `travis`. If people want to use a later version, they can in the future when we integrate custom services.

**note 3**: I'm not very familiar with `bluebird`, so some things might be done differently. Happy to receive feedback on that as well.

### CODE REVIEW

File: lib/parse-config.js [hunk 1]
```diff
@@ -20,5 +20,8 @@ module.exports = parseConfig;
  */
 
 function parseConfig (path) {
-	return yaml.parse(fs.readFileSync(path, 'utf-8'));
+	var config = yaml.parse(fs.readFileSync(path, 'utf-8'));
+	config.services = [].concat(config.services || []);
```
Comment (vdemedes) [inline lib/parse-config.js]:
```js
if (!config.services) {
  config.services = [];
}
```

File: index.js [hunk 2]
```diff
@@ -92,11 +94,21 @@ if (config.language !== 'node_js') {
...
+
+		return version;
+	})
+	.tap(function () {
+		return Promise.resolve(config.services).map(service.pull);
+	})
+	.map(function (version) {
 		var context = {
-			version: version,
+			version: format('%s-onbuild', version),
```
Comment (vdemedes) [inline index.js]: What is the reason behind this change?

File: lib/build.js [hunk 3]
```diff
@@ -23,7 +23,7 @@ module.exports = build;
  */
 
 function build (context) {
-	var dockerfile = format('FROM node:%s-onbuild', context.version);
+	var dockerfile = format('FROM node:%s', context.version);
```
Comment (vdemedes) [inline lib/build.js]: What is the reason behind this change?

File: lib/clean.js [hunk 4]
```diff
@@ -28,5 +28,16 @@ function clean (context) {
 
 	var image = format('test-%s-%s', context.name, context.version);
 
-	return run('docker', ['rmi', image]).return(context);
+	return Promise.resolve(context.services)
+		.map(function (service) {
+			var serviceName = format('%s-%s-%s', service, context.name, context.version);
+
+			return run('docker', ['kill', serviceName]).then(function () {
```
Comment (vdemedes) [inline lib/clean.js]: These should be parts of the promise chain, e.g.:

```js
.map()
.map(kill)
.map(rm)
```

File: lib/service.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+'use strict';
+
+/**
+ * Dependencies
+ */
+
+var format = require('util').format;
+var objectAssign = require('object-assign');
```
Comment (vdemedes) [inline lib/service.js]: `assign` instead of `objectAssign`, and ideally:

```js
var assign = Object.assign || require('object-assign');
```

File: lib/service.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+var run = require('../util/run');
+
+
+/**
+ * Services
+ */
+
+var services = {
+	mongodb: {
+		host: 'db',
```
Comment (vdemedes) [inline lib/service.js]: what is `host` for?

File: lib/service.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		tag: 'latest'
+	}
+};
+
+
+/**
+ * Expose
+ */
+
+module.exports.start = startService;
```
Comment (vdemedes) [inline lib/service.js]:
```js
exports.start = startService;
```

same for `pull`

File: lib/service.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+ * Start the service for a specific context
+ */
+
+function startService (serviceName, context) {
+	var service = services[serviceName];
+
+	var name = format('%s-%s-%s', serviceName, context.name, context.version);
+	var image = format('%s:%s', service.repository, service.tag);
+
+	return run('docker', ['run', '-d', '--name', name, image]).return(objectAssign(service, { name: serviceName }));
```
Comment (vdemedes) [inline lib/service.js]:
```js
var result = assign(service, { name: serviceName });
return run().return(result);
```

would be cleaner

File: lib/update-state.js [hunk 9]
```diff
@@ -30,6 +29,11 @@ function updateState (state) {
 
 		var currentState = state[version];
 
+		if (currentState === states.preparing) {
+			icon = chalk.grey(figures.circleDotted);
+			message = chalk.grey('preparing');
```
Comment (vdemedes) [inline lib/update-state.js]: "preparing" does not clearly explain what is going on, "starting services" would be better.

**Comment by vdemedes** [comment]
> note 1: travis exposes mongodb at the 127.0.0.1. This seems not possible when linking containers, but not sure, maybe there is a way. So at the moment, mongodb is exposed as db host.

Good point.

I added comments to your PR, otherwise everything looks good to me!

File: index.js (hunk 2 above)
Comment (SamVerschueren) [inline index.js]: Because I wanted to make `pullImage` reusable for my services. The `-onbuild` was hardcoded there.

I guess we should refactor that part to make it somewhat like the services. Create a map of supported languages with settings (like version, repository, ...). That will remove the `-onbuild` here.

File: lib/service.js (hunk 5 above)
Comment (SamVerschueren) [inline lib/service.js]: It already is a ponyfill :). https://github.com/sindresorhus/object-assign/blob/master/index.js

File: lib/service.js (hunk 6 above)
Comment (SamVerschueren) [inline lib/service.js]: The hostname that should be used in the code to connect with mongodb. It is the alias in the `--link` flag.

```
mongo.connect('mongodb://db/test');
```

File: lib/service.js (hunk 5 above)
Comment (vdemedes) [inline lib/service.js]: Oooh, thanks for the tip!
